spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -0,0 +1,268 @@
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.table = table;
+    this.spec = table.spec();
+    this.filter = Expressions.alwaysTrue();
+
+    long splitSize = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_SIZE,
+        TableProperties.SPLIT_SIZE_DEFAULT);
+    long targetFileSize = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES,
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT);
+    this.rewriteDataFileSizeInBytes = Math.min(splitSize, targetFileSize);
+
+    this.rewriteDataFileSizeInBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_SIZE,
+        TableProperties.SPLIT_SIZE_DEFAULT);
+    this.splitLookback = PropertyUtil.propertyAsInt(
+        table.properties(),
+        TableProperties.SPLIT_LOOKBACK,
+        TableProperties.SPLIT_LOOKBACK_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      this.fileIO = new HadoopFileIO(conf::value);
+    } else {
+      this.fileIO = table.io();
+    }
+    this.encryptionManager = table.encryption();
+  }
+
+  @Override
+  protected RewriteDataFilesAction self() {
+    return this;
+  }
+
+  @Override
+  protected Table table() {
+    return table;
+  }
+
+  /**
+   * Pass a PartitionSepc id to specify which PartitionSpec should be used in DataFile rewrite
+   *
+   * @param specId PartitionSpec id to rewrite
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction outputSpecId(int specId) {
+    Preconditions.checkArgument(table.specs().containsKey(specId), "Invalid spec id %d", specId);
+    this.spec = table.specs().get(specId);
+    return this;
+  }
+
+  /**
+   * Specify the rewrite data file size in bytes
+   *
+   * @param rewriteDataFileSize size of rewrite data file
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction rewriteDataFileSizeInBytes(long rewriteDataFileSize) {
+    Preconditions.checkArgument(rewriteDataFileSize > 0L, "Invalid rewrite data file size in bytes %d",
+        rewriteDataFileSize);
+    this.rewriteDataFileSizeInBytes = rewriteDataFileSize;
+    return this;
+  }
+
+  /**
+   * Specify the split lookback
+   *
+   * @param lookback lookback number to split
+   * @return this for method chaining
+   */
+  public RewriteDataFilesAction splitLookback(int lookback) {
+    Preconditions.checkArgument(lookback > 0L, "Invalid split lookback %d", lookback);
+    this.splitLookback = lookback;
+    return this;
+  }
+
+  /**
+   * Pass a row Expression to filter DataFiles to be rewritten. This is typically used when only rewriting DatFiles
+   * under some partitions.
```

Comment:
I think it is important to mention that all files that may contain data matching the filter may be rewritten.

When planning, the filter is used to match files inclusively: if a file may contain a matching row, the file matches. When matching file stats, the same inclusive approach is used. So if a user has a bucked table, `bucket(id, 16)` and asks to rewrite where `id = X`, then this will match any data file in `id_bucket = bucket(X, 16)`.